File: android/src/main/kotlin/com/loopit/minis/camera/CameraSession.kt
package com.loopit.minis.camera

import java.util.concurrent.atomic.AtomicReference

/**
 * Session state machine: idle → preview → recording → paused → stopped → finalized.
 *
 * Transitions are guarded; illegal transitions return false. Listeners are notified
 * on the calling thread (callers can hop threads if needed).
 */
class CameraSession {
    enum class State { IDLE, PREVIEW, RECORDING, PAUSED, STOPPED, FINALIZED, ERROR }

    fun interface Listener {
        fun onStateChanged(prev: State, next: State, code: String?, message: String?)
    }

    private val current = AtomicReference(State.IDLE)
    private val listeners = mutableListOf<Listener>()

    val state: State get() = current.get()

    fun addListener(l: Listener) { synchronized(listeners) { listeners.add(l) } }
    fun removeListener(l: Listener) { synchronized(listeners) { listeners.remove(l) } }

    @Synchronized
    fun transition(next: State, code: String? = null, message: String? = null): Boolean {
        val prev = current.get()
        if (!isLegal(prev, next)) return false
        current.set(next)
        val snapshot = synchronized(listeners) { listeners.toList() }
        snapshot.forEach { it.onStateChanged(prev, next, code, message) }
        return true
    }

    fun error(code: String, message: String?) {
        val prev = current.get()
        current.set(State.ERROR)
        val snapshot = synchronized(listeners) { listeners.toList() }
        snapshot.forEach { it.onStateChanged(prev, State.ERROR, code, message) }
    }

    /** Emit an out-of-band info event (e.g. timeLapseFinalized) without state transition. */
    fun emitInfo(code: String, message: String?) {
        val cur = current.get()
        val snapshot = synchronized(listeners) { listeners.toList() }
        snapshot.forEach { it.onStateChanged(cur, cur, code, message) }
    }

    private fun isLegal(prev: State, next: State): Boolean = when (prev) {
        State.IDLE -> next == State.PREVIEW || next == State.ERROR
        State.PREVIEW -> next == State.RECORDING || next == State.IDLE || next == State.ERROR
        State.RECORDING -> next == State.PAUSED || next == State.STOPPED || next == State.ERROR
        State.PAUSED -> next == State.RECORDING || next == State.STOPPED || next == State.ERROR
        State.STOPPED -> next == State.FINALIZED || next == State.PREVIEW || next == State.ERROR
        State.FINALIZED -> next == State.PREVIEW || next == State.IDLE
        State.ERROR -> next == State.IDLE || next == State.PREVIEW
    }
}

File: ios/Classes/Camera/CameraSession.swift
import Foundation

/// Minis iOS camera state machine, mirroring the Android version.
public enum MinisCameraState: String {
    case idle, preview, recording, paused, stopped, finalized, error
}

public protocol MinisCameraSessionListener: AnyObject {
    func cameraSessionDidTransition(
        from prev: MinisCameraState,
        to next: MinisCameraState,
        code: String?,
        message: String?
    )
}

public final class MinisCameraSessionMachine {
    public private(set) var state: MinisCameraState = .idle
    public weak var listener: MinisCameraSessionListener?
    private let queue = DispatchQueue(label: "minis.camera.session")

    public init() {}

    @discardableResult
    public func transition(to next: MinisCameraState, code: String? = nil, message: String? = nil) -> Bool {
        var ok = false
        queue.sync {
            if isLegal(prev: state, next: next) {
                let prev = state
                state = next
                ok = true
                listener?.cameraSessionDidTransition(from: prev, to: next, code: code, message: message)
            }
        }
        return ok
    }

    public func error(code: String, message: String?) {
        queue.sync {
            let prev = state
            state = .error
            listener?.cameraSessionDidTransition(from: prev, to: .error, code: code, message: message)
        }
    }

    private func isLegal(prev: MinisCameraState, next: MinisCameraState) -> Bool {
        switch prev {
        case .idle: return next == .preview || next == .error
        case .preview: return next == .recording || next == .idle || next == .error
        case .recording: return next == .paused || next == .stopped || next == .error
        case .paused: return next == .recording || next == .stopped || next == .error
        case .stopped: return next == .finalized || next == .preview || next == .error
        case .finalized: return next == .preview || next == .idle
        case .error: return next == .idle || next == .preview
        }
    }
}
